libnetwork: processSetKeyReexec() remove defer(), and some refactoring

thaJeztah opened the pull request:
- [x] follow-up to https://github.com/moby/moby/pull/43505

### libnetwork: processSetKeyReexec() remove defer()

Split the function into a "backing" function that returns an error, and the
re-exec entrypoint, which handles the error to provide a more idiomatic approach.

This was part of a larger change accross multiple re-exec functions (now removed).

For history's sake; here's the description for that;

> The `reexec.Register()` function accepts reexec entrypoints, which are a `func()`
> without return (matching a binary's `main()` function). As these functions cannot
> return an error, it's the entrypoint's responsibility to handle any error, and to
> indicate failures through `os.Exit()`.
> 
> I noticed that some of these entrypoint functions had `defer()` statements, but
> called `os.Exit()` either explicitly or implicitly (e.g. through `logrus.Fatal()`).
> defer statements are not executed if `os.Exit()` is called, which rendered these
> statements useless.
> 
> While I doubt these were problematic (I expect files to be closed when the process
> exists, and `runtime.LockOSThread()` to not have side-effects after exit), it also
> didn't seem to "hurt" to call these as was expected by the function.
> 
> This patch rewrites some of the entrypoints to split them into a "backing function"
> that can return an error (being slightly more iodiomatic Go) and an wrapper function
> to act as entrypoint (which can handle the error and exit the executable).
> 
> To some extend, I'm wondering if we should change the signatures of the entrypoints
> to return an error so that `reexec.Init()` can handle (or return) the errors, so
> that logging can be handled more consistently (currently, some some use logrus,
> some just print); this would also keep logging out of some packages, as well as
> allows us to provide more metadata about the error (which reexec produced the
> error for example).
> 
> A quick search showed that there's some external consumers of pkg/reexec, so I
> kept this for a future discussion / exercise.


### libnetwork: processSetKeyReexec: don't use logrus.Fatal()

Just print the error and os.Exit() instead, which makes it more
explicit that we're exiting, and there's no need to decorate the
error.

### libnetwork: don't register "libnetwork-setkey" re-exec on non-unix

It's a no-op on Windows and other non-Linux, non-FreeBSD platforms,
so there's no need to register the re-exec.

### libnetwork: setKey(): remove intermediate buffer

Use `json.NewDecoder()` instead, which accepts a reader.

### libnetwork: inline sendKey() into SetExternalKey()

This function included a defer to close the `net.Conn` if an error occurred,
but the calling function (`SetExternalKey()`) also had a defer to close it
unconditionally.

Rewrite it to use `json.NewEncoder()`, which accepts a writer, and inline
the code.


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
libnetwork: processSetKeyReexec() remove defer()

Split the function into a "backing" function that returns an error, and the
re-exec entrypoint, which handles the error to provide a more idiomatic approach.

This was part of a larger change accross multiple re-exec functions (now removed).

For history's sake; here's the description for that;

The `reexec.Register()` function accepts reexec entrypoints, which are a `func()`
without return (matching a binary's `main()` function). As these functions cannot
return an error, it's the entrypoint's responsibility to handle any error, and to
indicate failures through `os.Exit()`.

I noticed that some of these entrypoint functions had `defer()` statements, but
called `os.Exit()` either explicitly or implicitly (e.g. through `logrus.Fatal()`).
defer statements are not executed if `os.Exit()` is called, which rendered these
statements useless.

While I doubt these were problematic (I expect files to be closed when the process
exists, and `runtime.LockOSThread()` to not have side-effects after exit), it also
didn't seem to "hurt" to call these as was expected by the function.

This patch rewrites some of the entrypoints to split them into a "backing function"
that can return an error (being slightly more iodiomatic Go) and an wrapper function
to act as entrypoint (which can handle the error and exit the executable).

To some extend, I'm wondering if we should change the signatures of the entrypoints
to return an error so that `reexec.Init()` can handle (or return) the errors, so
that logging can be handled more consistently (currently, some some use logrus,
some just print); this would also keep logging out of some packages, as well as
allows us to provide more metadata about the error (which reexec produced the
error for example).

A quick search showed that there's some external consumers of pkg/reexec, so I
kept this for a future discussion / exercise.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
These look legit; but will have to look what's causing the failure;

```
Running /go/src/github.com/docker/docker/integration/plugin/graphdriver (amd64.integration.plugin.graphdriver) flags=-test.v -test.timeout=120m 
 INFO: Testing against a local daemon
 === RUN   TestExternalGraphDriver
 === RUN   TestExternalGraphDriver/json
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest
 === RUN   TestExternalGraphDriver/spec
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest
 === RUN   TestExternalGraphDriver/pull
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest@sha256:95cf004f559831017cdf4628aaf1bb30133677be8702a8c5f2994629f637a209
     panic.go:482: [db702f8451897] daemon is not started
 --- FAIL: TestExternalGraphDriver (1.43s)
     --- FAIL: TestExternalGraphDriver/json (0.63s)
     --- FAIL: TestExternalGraphDriver/spec (0.17s)
     --- FAIL: TestExternalGraphDriver/pull (0.61s)
 === RUN   TestGraphdriverPluginV2
 --- PASS: TestGraphdriverPluginV2 (3.88s)
 FAIL
 
 === Failed
 === FAIL: amd64.integration.plugin.graphdriver TestExternalGraphDriver/json (0.63s)
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest
     --- FAIL: TestExternalGraphDriver/json (0.63s)
 
 === FAIL: amd64.integration.plugin.graphdriver TestExternalGraphDriver/spec (0.17s)
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest
     --- FAIL: TestExternalGraphDriver/spec (0.17s)
 
 === FAIL: amd64.integration.plugin.graphdriver TestExternalGraphDriver/pull (0.61s)
     container.go:52: assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest@sha256:95cf004f559831017cdf4628aaf1bb30133677be8702a8c5f2994629f637a209
     panic.go:482: [db702f8451897] daemon is not started
     --- FAIL: TestExternalGraphDriver/pull (0.61s)
 
 === FAIL: amd64.integration.plugin.graphdriver TestExternalGraphDriver (1.43s)
```

samuelkarp reviewed (commented):
nit: typo in the commit message (s/To some extend/To some extent/).

Generally in favor of this type of change though.

tianon reviewed (approved):
I wonder if code linters will eventually start to catch issues like this :eyes:

thaJeztah commented:
> I wonder if code linters will eventually start to catch issues like this 👀

This one may be hard, as `logrus.Fatal()` can be overridden (default is os.Exit(), but it can be overridden)

Thanks for review, both! Erm, but CI is failing, so I still need to look into that one failure 😂 (see my comment earlier)

I'll ping if I had time to investigate that one ☺️

thaJeztah committed:
libnetwork: processSetKeyReexec: don't use logrus.Fatal()

Just print the error and os.Exit() instead, which makes it more
explicit that we're exiting, and there's no need to decorate the
error.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: don't register "libnetwork-setkey" re-exec on non-unix

It's a no-op on Windows and other non-Linux, non-FreeBSD platforms,
so there's no need to register the re-exec.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: setKey(): remove intermediate buffer

Use json.NewDecoder() instead, which accepts a reader.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: inline sendKey() into SetExternalKey()

This function included a defer to close the net.Conn if an error occurred,
but the calling function (SetExternalKey()) also had a defer to close it
unconditionally.

Rewrite it to use json.NewEncoder(), which accepts a writer, and inline
the code.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

corhere commented on libnetwork/sandbox_externalkey_unix.go:
```diff
@@ -24,82 +25,68 @@ const (
-		}
-	}()
+	if err := setKey(); err != nil {
+		println(err)
```
The bootstrapping `println` builtin [is not technically part of the Go language.](https://go.dev/ref/spec#Bootstrapping)
```suggestion
		fmt.Println(err)
```

thaJeztah replied:
Oh! Good call yes; that was originally a "WIP" commit, and I must've been lazy when I wrote that code (`println` prints to stderr, so saved me from having to use `os.StdOut` etc).

Let me update

corhere reviewed (commented):
Everything else LGTM

thaJeztah commented:
Fixed the `println()` ptal 👍

thaJeztah commented:
Failures are unrelated, and known

- https://github.com/moby/moby/pull/45415
